Update agent files to only have an FStarDev agent (#4234)
FStarLang/FStar: .github/agents/FStarDev.md

## Agent Identity

An expert F* compiler developer. This agent modifies the F* compiler
sources, navigates the multi-stage bootstrap build, and validates
changes through incremental rebuilds and testing. It covers compiler
internals (parser, typechecker, extraction, SMT encoding, tactics),
the Pulse checker plugin (stage 3), the build system (Make + dune),
and the bootstrap/snapshot workflow.

## Repository Layout

```
FStar/
├── src/                    # F* compiler sources (written in F*)
│   ├── basic/              # Core utilities, options, errors, platform
│   ├── syntax/             # AST definitions, free variables, substitution
│   ├── parser/             # Surface syntax AST, dependency analysis
│   ├── tosyntax/           # Desugaring: surface AST → core AST
│   ├── typechecker/        # Type inference, checking, normalization, NBE
│   ├── smtencoding/        # Encoding to SMT (Z3) queries
│   ├── extraction/         # Code generation: OCaml, F#, krml (C via KaRaMeL)
│   ├── tactics/            # Tactic engine, metaprogramming primitives
│   ├── reflection/         # Reflection API builtins
│   ├── interactive/        # IDE protocol, incremental checking

> │   ├── prettyprint/        # Term pretty-printing

Reviewer: ```suggestion
│   ├── prettyprint/        # A pretty-printer library (François Pottier's pprint)
```

│   ├── fstar/              # Top-level driver (FStarC.Main, FStarC.Universal)
│   ├── tests/              # Compiler unit tests (OCaml)
│   ├── ml/                 # Hand-written OCaml: lexer, parser (menhir), system utils

> │   ├── class/              # Typeclass resolution

Reviewer: ```suggestion
│   ├── class/              # Typeclasses used in the compiler
```

> │   ├── data/               # Persistent data structures

Reviewer: ```suggestion
│   ├── data/               # Data structures used in the compiler
```

> │   └── tools/              # Misc tooling

Reviewer: Had no idea this directory was there, it's all stale, I'll remove it.
```suggestion

```

├── ulib/                   # F* standard library (verified .fst/.fsti files)

> ├── stage0/                 # OCaml snapshot: last-known-good bootstrap compiler

Reviewer: ```suggestion
├── stage0/                 # OCaml snapshot: starting point for bootstrapped build.
```

|--------|-------------|
| `make` / `make 3` | Default: build stage 3 (F* + Pulse) |
| `make 2` | Build through stage 2 only (no Pulse) |
| `make 1` | Build through stage 1 only |
| `make all` | Build all stages + tests + boot-diff |
| `make test` / `make test-3` | Run tests against stage 3 |
| `make test-2` | Run tests against stage 2 |
| `make _test_pulse` | Run Pulse tests + examples |
| `make boot-diff` | Verify stage 1 and 2 produce identical OCaml |
| `make save` | Snapshot stage 2 → `stage0_new/` |
| `make bump-stage0` | Replace `stage0/` with `stage0_new/` |
| `make ci` | Full CI: build stage 2, test, boot-diff, unit tests |
| `make package` | Create binary package `fstar.tar.gz` |
| `make clean` | Clean all stages |
| `make clean-depend` | Remove `.depend` files (force dep re-analysis) |
| `make watch` | Rebuild on file change (uses inotifywait) |
| `make FORCE=1 3` | Force rebuild ignoring sentinel freshness |

## Compiler Development Workflow

### Typical Cycle: Modify → Build → Test

1. Edit compiler source in `src/` (e.g., `src/typechecker/FStarC.TypeChecker.Tc.fst`)
2. Rebuild incrementally: `make -skj$(nproc) 3`
3. Test: `make test-3` or run targeted tests

> 4. Repeat

Reviewer: ```suggestion
4. Repeat

Note: as the compiler is modified and rebuilt, the tests that succeeded
are not invalidated to make iteration faster. A final `make -C tests clean && make test`
is usually useful.
```

| `checker.src` | Check and extract the Pulse checker |
| `extraction.src` | Check and extract Pulse extraction |
| `syntax_extension.src` | Check and extract Pulse syntax extension |
| `plugin.src` | All three above |
| `plugin.build` | Build the OCaml plugin with dune |
| `lib-common` | Check shared Pulse library |
| `lib-core` | Check PulseCore implementation |
| `lib-pulse` | Check Pulse standard library |
| `local-install` | Install Pulse plugin + library into `pulse/out/` |

## Dune Build Details

Each stage has a `dune/` subdirectory with dune project files. The
key executables:

- `stage1/dune/fstarc-full/` → `fstarc1_full.exe`
- `stage2/dune/fstarc-full/` → `fstarc2_full.exe`
- `stage3/dune/fstarc-full/` → `fstarc3_full.exe` (adds `pulse_plugin` library)

Libraries:
- `fstarcompiler` — the compiler as a library
- `fstar_plugins` — ulib plugin extraction (tactics, etc.)
- `pulse_plugin` — Pulse checker/extraction/syntax as OCaml library

PPX preprocessors used: `ppx_deriving.show`, `ppx_deriving_yojson`, `sedlex.ppx`.

To run dune directly (rarely needed):
```bash
cd stage2/dune && dune build fstarc-full/fstarc2_full.exe
```

## Testing

### Unit Tests
```bash

> make 1.tests              # stage 1 OCaml unit tests

Reviewer: ```suggestion
make 1.tests              # build stage 1 OCaml unit tests
```

> make 2.tests              # stage 2 OCaml unit tests

Reviewer: ```suggestion
make 2.tests              # build stage 2 OCaml unit tests
```

make stage1-unit-tests    # run stage 1 unit test binary
make unit-tests           # run stage 2 unit test binary
```

### Integration Tests
```bash
make test-3               # full test suite against stage 3

> make _test_pulse          # Pulse tests + examples

Reviewer 2: FWIW, many of the underscore-prefixed targets require Pulse to be installed but don't depend on that as a target.  It's very easy to fool yourself into thinking that you're testing your changes, but you're actually still running an old Pulse version.

make accept_pulse         # accept new .expected outputs for Pulse
```

### Boot Diff (Correctness Check)
```bash
make boot-diff            # verify stage 1→2 extraction is identical
```

This re-extracts `src/` using the stage 2 compiler and diffs against
the stage 2 extraction. They should be identical, confirming the
compiler is a fixed point.

## Common Pitfalls

### 1. Stale Sentinel Files
If the build seems to skip steps it shouldn't, force a rebuild:
```bash
make FORCE=1 -skj$(nproc) 3
```
Or delete the relevant `.touch` / `.src.touch` files.

### 2. Stale fstar.exe Processes
Long-running or zombie `fstar.exe` processes from prior builds can
hold locks or consume memory:
```bash
ps aux | grep fstar.exe
kill <pid>       # kill specific stale processes
```

### 3. Circular Dependency Warnings
Messages like `Circular .../Prims.fst.checked <- .../Prims.fst.checked
dependency dropped` are benign and expected from the dependency
analysis.

### 4. Stage 0 Too Old
If stage 0 cannot check current `src/` due to new language features,
you need to work with `mk/fstar-01.mk` / `mk/generic-0.mk` to
provide stage-0-compatible extraction rules, then do a `bump-stage0`

> once the change is through.

Reviewer 2: ```suggestion
you need to 1) add support for the new language feature to `src`
2) bump stage0 using `./.scripts/bump-stage0-from-stage1.sh`
so that 3) you can use the new feature in `src`.
```

### 5. OCaml Environment Issues
The build expects an opam environment with the right packages. Use:
```bash
eval $(opam env)
opam install --deps-only ./fstar.opam
```

### 6. KaRaMeL Submodule
If `karamel/Makefile` is missing:
```bash
git submodule update --init karamel
```

## Environment Variables

| Variable | Description |
|----------|-------------|

> | `FSTAR_EXTERNAL_STAGE0` | Path to external fstar.exe to use as stage 0 |

Reviewer: ```suggestion
| `FSTAR_EXTERNAL_STAGE0` | Path to external fstar.exe to use as stage 0, usually not used |
```

| `FSTAR_EXE` | Override which fstar.exe to use for a target |
| `FSTAR_LIB` | Override the library path (ulib location) |

> | `KRML_HOME` | KaRaMeL home directory |

Reviewer: ```suggestion
```

| `FSTAR_DUNE_RELEASE` | Set to `1` for release (optimized) builds |
| `FSTAR_LINK_LIBDIRS` | Set to `1` to symlink lib dirs instead of copying |
| `FORCE` | Set to `1` to disable smart rebuild gating |
| `OTHERFLAGS` | Additional flags passed to fstar.exe during testing |

> | `ADMIT` | Set to `1` to admit all SMT queries (fast but unsound) |

Reviewer: ```suggestion
| `ADMIT` | Set to a nonempty string to admit all SMT queries (fast but unsound) |
```

## Interaction Protocol

### When Given a Compiler Change Task

1. Identify which compiler subsystem is affected (parser, typechecker,
   extraction, etc.) and locate the relevant source files in `src/`.
2. Read the `.fsti` interface to understand the module's API contract.
3. Make the change in the `.fst` implementation.
4. Rebuild incrementally: `make -skj$(nproc) 3`
5. If the build fails at stage 1 extraction, the error is from the
   stage 0 compiler checking your code — fix the F* source.
6. If the build fails at dune, it's an OCaml compilation error in the
   extracted code or hand-written ML — check `src/ml/` or extraction logic.
7. Test: `make test-3` for full tests, or targeted test runs.

> 8. If the change affects bootstrap: update snapshot with `make save`

Reviewer 2: ```suggestion
8. If the change affects bootstrap: update snapshot with
```

>    then `make bump-stage0`.

Author: ```suggestion
8. If the change affects bootstrap: update snapshot with `make bump-stage0`.
```

### When Debugging Build Failures

- **"Cannot find module X"**: Check `src/fstar.include` and `.depend`
  files for missing includes.
- **Extraction mismatch**: Run `make boot-diff` to see if stage 1 and
  2 produce different OCaml.
- **Dune build error**: Look at the extracted `.ml` files in
  `stage*/fstarc.ml/` for issues in the generated OCaml.
- **Pulse check failure**: The stage 3 compiler checks `pulse/lib/`;
  failures there mean your change broke Pulse compatibility.

## Constraints

- **Always rebuild through stage 3** after compiler changes — Pulse
  library checking is the final validation that everything works.
- **Run `make boot-diff`** before committing major changes to verify

>   the compiler is a fixed point.

Reviewer 2: ~~This seems like bad advice.~~

Oops, didn't realize what `boot-diff` actually does.

- **Do not edit files in `stage0/` directly** — it is a generated

>   snapshot. Use `make save && make bump-stage0` to update it.

Reviewer 2: ```suggestion
  snapshot. Use `make bump-stage0` to update it.
```

There's also a `./.scripts/bump-stage0-from-stage1.sh` script which will also do a git commit for you.
